Assignment1 PR

jeki01 opened the pull request.

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,116 @@
+        return num;
+    }
+    // if expression inside brackets
+    else if (**expr == '(') {
```
It's good that we are handling the brackets here, but as it is mentioned in problem statement that `The input string contains integers and operators (+, -, *, /) only.` , We should treat brackets as invalid expression,

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,116 @@
+
+// factor: handles numbers and brackets
+int solveFactor(const char **expr, int *err) {
+    skipSpaces(expr);
```
Currently we are calling this function at multiple locations because we are passing whole expression as a function argument. Instead, we can create a loop in the `solveExpr` function and call function inside loop, so we don't have to worry about calling it every time.

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,116 @@
+        return val;
+    }
+    else {
+        *err = 1; // invalid input
```
This will only work if user gives invalid character at starting of the expression. If there is invalid character in middle of expression, then our code will misbehave. For example: Try running program with `4#5-3`. 
You need to validate expression in some other way.

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,116 @@
+        int right = solveFactor(expr, err);
+
+        if (op == '*') {
+            result *= right;
```
Our code can display incorrect value to the user as we are not handling the range overflow here. For example, 1000000*1000000 will be out of int capacity on both `32 and 64 bit` systems, which would result in undefined result value. This applies to all type of operations not only multiplication.

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,116 @@
+    char input[100];
+
+    printf("Enter expression: ");
+    if (!fgets(input, sizeof(input), stdin)) {
```
User may get incorrect result when expression given is of longer size than our maximum input buffer length (100 in our case). To handle this, we should have a logic to display error if expression is longer than expected.

jeki01 commented:
brackets are now invalid, any wrong character gives an error, division by zero is handled, integer overflow is checked, and inputs longer than 99 characters are not allowed.

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,158 @@
+            break;
+        default: return 0;
+    }
+    return (result > INT_MAX || result < INT_MIN);
```
This is not the correct way to handle the overflow.

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,158 @@
+
+// factor: handles numbers only (brackets invalid)
+int solveFactor(int *err) {
+    skipSpaces();
```
Comment about these function calls is not addressed yet.

jeki01 commented:
Updated willOverflow() to check for overflow mathematically before performing operations.,Updated solveFactor() function comment to clearly explain its purpose

Bhavik-Ubhadiya commented on Assignment1.c:
```diff
@@ -0,0 +1,170 @@
+
+// It skips any leading spaces and checks the current character.
+int solveFactor(int *err) {
+    skipSpaces();
```
I am approving this for now, but my suggestion is still not addressed. You can go through my first comment on these `skipSpaces()` calls again and try to work on it on local. Let me know if you have any doubt on that.
